EasyPost/Services/ShipmentService.cs

```diff
@@ -0,0 +1,129 @@
+        ///     All invalid keys will be ignored.
+        /// </param>
+        /// <param name="withCarbonOffset">Whether to use carbon offset when creating the shipment.</param>
+        /// <returns>An EasyPost.Shipment instance.</returns>
+        [CrudOperations.Create]
+        public async Task<Shipment> Create(Dictionary<string, object> parameters, bool withCarbonOffset = false)
+        {
+            parameters = parameters.Wrap("shipment");
+            parameters.Add("carbon_offset", withCarbonOffset);
+            return await Create<Shipment>("shipments", parameters);
+        }
+
+        /// <summary>
+        ///     Create a return shipment for a shipment.
+        /// </summary>
+        /// <param name="shipment">Shipment to return.</param>
+        /// <returns>A return shipment.</returns>
+        [CrudOperations.Create]
+        public async Task<Shipment> CreateReturn(Shipment shipment)
+        {
+            if (shipment.id == null)
+            {
+                throw new Exception("Shipment.id is null.");
+            }
+
+            return await Create(
+                new Dictionary<string, object>
+                {
+                    {
+                        "to_address", shipment.from_address
+                    },
+                    {
+                        "from_address", shipment.to_address
+                    },
+                    {
+                        "parcel", shipment.parcel
+                    },
+                    {
+                        "is_return", true
+                    },
+                });
+        }
+
+        /// <summary>
+        ///     Get a paginated list of shipments.
+        /// </summary>
+        /// <param name="parameters">
+        ///     Optional dictionary containing parameters to filter the list with. Valid pairs:
+        ///     * {"before_id", string} String representing a Shipment. Starts with "shp_". Only retrieve shipments created before
+        ///     this id. Takes precedence over after_id.
+        ///     * {"after_id", string} String representing a Shipment. Starts with "shp_". Only retrieve shipments created after
+        ///     this id.
+        ///     * {"start_datetime", DateTime} Starting time for the search.
+        ///     * {"end_datetime", DateTime} Ending time for the search.
+        ///     * {"page_size", int} Size of page. Default to 20.
+        ///     * {"purchased", bool} If true only display purchased shipments.
+        ///     All invalid keys will be ignored.
+        /// </param>
+        /// <returns>An EasyPost.ShipmentCollection instance.</returns>
+        [CrudOperations.Read]
+        public async Task<ShipmentCollection> All(Dictionary<string, object>? parameters = null)
+        {
+            ShipmentCollection shipmentCollection = await List<ShipmentCollection>("shipments", parameters);
+            shipmentCollection.Client = Client;
```

Comment:
Question: what does this line do? I have seen this in several places for `All` function

Reply:
Line 99 makes the API call and deserializes the JSON data into a `ShipmentCollection` object (a wrapper that stores a list of shipments as well as information on `has_more`, etc.)

Line 100 copies the API client that was used to make this call into the collection, so the collection object itself can make API calls via instance methods.